### PR TITLE
Simplify and fix colspan and rowspan filling in html_table

### PR DESCRIPTION
**Pull request by jaanos**
When reading a HTML table containing cells spanning multiple rows, `html_table` may behave incorrectly - I have set up a [repository](https://github.com/jaanos/rvest-table-fill) illustrating the issue. This pull request fixes the problem, at least for well-formed tables:

* If not all `<tr>`s span the same number of columns, the total number of cells is compared to the product of the number of rows and the maximal span to determine whether `fill = TRUE` is needed. The current checks (comparing against the sums of rowspans and colspans) never succeed.
* When filling the matrix, vectors are maintained indicating how many cells to the right and below the current cell spans. This allows skipping the cells spanning from previous rows, thus emulating how browsers display tables.

This should work for all tables in which the total number of cells in each row (including cells from previous rows spanning down to the current row) is constant, no cell spans below the last `<tr>`, and cells do not overlap (i.e., multicolumn cells do not span over a range of columns containing multirow cells spanning from previous rows).

* A malformed table may have more columns than the maximal sum of colspans in a `<tr>` (in a well-formed table, the maximum is achieved in the first row). This patch silently drops any cells that would fall outside the output matrix. If issuing a warning or even an error would be preferable, I can change that.
* Similarly, any multirow cells spanning below the last `<tr>` will be cut at the bottom of the table (but here, no data loss can occur).
* If there are overlapping cells, data from the cell appearing later in the table (i.e., beginning in a row closer to the bottom and in a column closer to the left) is used at points of overlap.
* It may happen that a malformed table passes the colspan or number of cells test even with `fill = FALSE`, thus keeping `NA`s in shorter rows.

Detecting the actual number of columns would of course be possible, but would require another pass through the cells (or adding columns on-the-fly), which might slow down the function.

Anyway, using this patch, the table from the repository linked above can be read correctly without using `fill = TRUE`.

### CODE REVIEW

**Comment by jaanos** [comment]
The last commit still fails with one test, but it is that test that is incorrectly written. Should I correct it?

**Comment by jaanos** [comment]
Any chance this PR is considered? I have added a test from #236, showing that this solution generalizes the one given there.

**Comment by hadley** [comment]
Sorry for the long delay. Are you still interested in finishing this off?

**Comment by jaanos** [comment]
Sure, I am still interested.

File: tests/testthat/test-tables.R [hunk 1]
```diff
@@ -258,3 +258,76 @@ test_that("a correct but slightly pathological table correctly parsed", {
                c(rep('Categorical Columns', 2),
                  rep('Numerical Columns', 2)))
 })
+
+test_that("Test rowspans are respected when fill=True", {
+  select <- minimal_html("table parsing", '
```
Comment (hadley) [inline tests/testthat/test-tables.R:263]: I wonder if it might be better to break this down into a few simpler tables that each illustrate one problem?

**Review by hadley — COMMENTED** [review]
Thanks for working on this!

File: R/table.R [hunk 2]
```diff
@@ -89,10 +89,11 @@ html_table.xml_node <- function(x, header = NA, trim = TRUE,
   p <- unique(vapply(ncols, sum, integer(1)))
   maxp <- max(p)
 
-  if (length(p) > 1 & maxp * n != sum(unlist(nrows)) &
-      maxp * n != sum(unlist(ncols))) {
+  if (length(p) > 1) {
     # then malformed table is not parsable by smart filling solution
-    if (!fill) { # fill must then be specified to allow filling with NAs
+    ncells <- sum(unlist(lapply(seq_len(n),
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/table.R:94]: Or maybe it would be better to make `ncols` and `nrows` vectors rather than lists.

[the remark this answers is not in the document]
Comment (hadley) [reply R/table.R:94]: Oh they're lists by row

**Comment by hadley** [comment]
I think I see a simpler way to attack this - let me have a go tomorrow.

**Comment by jaanos** [comment]
Sure, go ahead. Sorry for my non-responsiveness, I am a bit busy these days.

**Comment by hadley** [comment]
Total rewrite in #293 — I think this should work reliably with any table, because it follows the same principles as a browser.

**Comment by jaanos** [comment]
Thanks, works correctly now!

**Pull request closed without merging**